### PR TITLE
feature/length-padding

### PR DESCRIPTION
**Pull request by simylein**
This is my first attempt at implementing #13 
I am particularly unsure if I am supposed to pass command line args like that. 
And yes, I know I called it `--auto-paddding` instead of `--skip-length-validation-and-pad`. 
I think the latter is just too long as a command line flag (also having trouble finding a short version of it)
So therefore I propose `--auto-padding` or `--payload-padding` or something similar. 
However, if you insist on your flag name I will gladly add another commit renaming it. 
Please let me know if this is the right approach, or if you would like to have it done differently (coding wise).

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -40,6 +41,12 @@ func init() {
 	if err != nil {
 		slog.Error("error while binding json flag", slog.Any("error", err))
 	}
+
+	rootCmd.PersistentFlags().BoolVarP(&AutoPadding, "auto-padding", "p", false, "Enable automatic padding of payload. (default: false)")
```
Comment (michaelbeutler) [inline cmd/root.go:45]: Please add a warning that the output may get corrupted.

**Comment by michaelbeutler** [comment]
I would still suggest the flag containing some kind of trigger word like "skip" or "unsafe". This flag is only a workaround for very unique projects.

**Review by michaelbeutler — CHANGES_REQUESTED** [review]
Please add more tests to keep the test coverage.

File: cmd/root.go (hunk 1 above)
Comment (simylein) [reply cmd/root.go]: Screenshot 2024-12-04 at 14 56 37
like this?

**Comment by simylein** [comment]
We kinda have a problem, because there are optional fields. 
Consider a message of 12 bytes which uses a port config on which the first 8 bytes are required and a further 8 bytes optional. 
Now if the message is 11 bytes long there is no way of knowing how many zeros to add because it could be just 1 byte of zeros or also 5 bytes of zeros. 
We can just say to add zeros until the nearest full byte, but then your example of `123` -> `00000123` is impossible as we need to add 2 bytes and 4 bits for which we need to know that the message needs to be 4 bytes long.
I am not really sure how to proceed, I can just add test cases which check for scrambled data, or only test port implementations which do not have any optional fields.

File: cmd/root.go (hunk 1 above)
Comment (michaelbeutler) [reply cmd/root.go]: Looks good thanks!

**Comment by michaelbeutler** [comment]
Since this is an workaround, we don't expect this to happen and no optional fields will be set when this flag is active.

**Comment by simylein** [comment]
> Please add more tests to keep the test coverage.

Has been done, thanks for the reminder 🙂

**Comment by simylein** [comment]
> I would still suggest the flag containing some kind of trigger word like "skip" or "unsafe". This flag is only a workaround for very unique projects.

What do we do about the naming situation? Would you like me to rename to `--skip-length-validation-and-pad`?

**Comment by michaelbeutler** [comment]
We can go with `auto-padding` but should remove the shortcut `-p`. Additionally, we should consider using the Functional Options Pattern to minimize interface changes.
See example: https://dev.to/kittipat1413/understanding-the-options-pattern-in-go-390c

````go
type Option func(*Decoder)

func WithAutoPadding() Option {
    return func(d *Decoder) {
        d.autoPadding = true
    }
}

**Comment by simylein** [comment]
> We can go with `auto-padding` but should remove the shortcut `-p`. Additionally, we should consider using the Functional Options Pattern to minimize interface changes. See example: https://dev.to/kittipat1413/understanding-the-options-pattern-in-go-390c
> 
> ```go
> type Option func(*Decoder)
> 
> func WithAutoPadding() Option {
>     return func(d *Decoder) {
>         d.autoPadding = true
>     }
> }
> ```

I would need some guidance with that as I looked into it and couldn't quite figure out how to apply this in the code base.

**Comment by simylein** [comment]
thank you for implementing the functional options

**Review by michaelbeutler — APPROVED** [review]

**Pull request merged**